# Dataquest Curriculum Project 3 - Exploring Ebay Car Sales Data

The focus of this project is light data cleaning.

We will be working with a dataset of used cars listed on *eBay Kleinazeigen*, a classifieds section of the German eBay website. The data were originally scraped by [Orges Leka on Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data) 2 years ago, and the data here are 50,000 data points sampled from it. These are the fields:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

In [112]:
%%javascript
$.getScript('https://kmahelona.github.io/ipython_notebook_goodies/ipython_notebook_toc.js')

<IPython.core.display.Javascript object>

<h1 id="tocheading">Table of Contents</h1>
<div id="toc"></div>

# Preview of the data

In [113]:
import numpy as np
import pandas as pd

try:
    autos = pd.read_csv("autos.csv")
except:
    try:
        autos = pd.read_csv("autos.csv", encoding="Latin-1")
    except:
        autos = pd.read_csv("autos.csv", encoding="Windows-1252")

In [114]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [115]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [116]:
autos.tail()

dateCrawled                                               name  \
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

      brand notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
49995  audi              nein  2016-03-27 00:00:00             0       82131   
49996  opel              nein  2016-03-28 00:00:00             0       44807   
49997  fiat              nein  2016-04-02 00:00:00             0       73430   
49998  audi              nein  2016-03-08 00:00:00             0       35683   
49999  opel              nein  2016-03-13 00:00:00             0       45897   

                  lastSeen  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

In [117]:
# gives the counts of null values for each column
autos.isnull().sum()

dateCrawled               0
name                      0
seller                    0
offerType                 0
price                     0
abtest                    0
vehicleType            5095
yearOfRegistration        0
gearbox                2680
powerPS                   0
model                  2758
odometer                  0
monthOfRegistration       0
fuelType               4482
brand                     0
notRepairedDamage      9829
dateCreated               0
nrOfPictures              0
postalCode                0
lastSeen                  0
dtype: int64

*We may later wish to address the fact that vehicleType, gearbox, model, fuelType, and/or notRepairedDamage have null values.*

# Renaming column labels to snakecase

In [118]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [119]:
new_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.columns = new_names
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

*We just changed the column labels to snakecase because that's preferred in Python*

# Searching for data cleaning tasks

In [120]:
# descriptive statistics (include object/string columns)
autos.describe(include=['O'])

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   

       vehicle_type  gearbox  model   odometer fuel_type       brand  \
count         44905    47320  47242      50000     45518       50000   
unique            8        2    245         13         7          40   
top       limousine  manuell   golf  150,000km    benzin  volkswagen   
freq          12859    36993   4024      32424     30107       10687   

       unrepaired_damage           ad_created            last_seen  
count              40171                50000                50000  
unique                 2                   76                39481  
top                 nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq               35232                 1946                    8

Columns that are candidates to drop:
* `seller` - for practically every single row this is "privat"
* `offer_type` - for practically every single row this is "Angebot"

Columns that need more investigation:
* `price` - The most frequent price is $0. That doesn't seem right.

Columns that need to be cleaned / converted to numeric data type:
* `price` - remove "$"
* `odometer` - remove "km"

## Dropping seller & offer type

In [121]:
# dropping
autos = autos.drop(labels=["seller", "offer_type"], axis=1)
autos.head(1)

date_crawled                              name   price  ab_test  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  $5,000  control   

  vehicle_type  registration_year  gearbox  power_PS   model   odometer  \
0          bus               2004  manuell       158  andere  150,000km   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54

## Investigating rows with price as 0

In [122]:
# investigating rows with price as $0
zero_price_bool = autos["price"] == "$0"
autos.loc[zero_price_bool].head(10)

date_crawled                                               name  \
27   2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71   2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80   2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87   2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99   2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118  2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146  2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167  2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180  2016-03-19 10:50:25                                       Zu_verkaufen   
226  2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   

    price  ab_test vehicle_type  registration_year  gearbox  power_PS  \
27     $0  control          NaN               2005      NaN         0   
71     $0  control          NaN               1990  manuell         0   
80     $0  control        coupe               1999  manuell        99   
87     $0  control          NaN               2000      NaN         0   
99     $0  control       cabrio               2011  manuell         0   
118    $0  control          bus               2001  manuell       204   
146    $0     test   kleinwagen               1996  manuell        75   
167    $0  control          NaN               2011      NaN         0   
180    $0     test          NaN               2016  manuell        98   
226    $0  control       cabrio               1967  manuell       160   

           model   odometer  registration_month fuel_type       brand  \
27           NaN  150,000km                   0       NaN        ford   
71           NaN    5,000km                   0    benzin        opel   
80       primera  150,000km                   3    benzin      nissan   
87           5er  150,000km                   0       NaN         bmw   
99       2_reihe   60,000km                   7    diesel     peugeot   
118       sharan  150,000km                   7    benzin  volkswagen   
146       fiesta   20,000km                   8    benzin        ford   
167  transporter    5,000km                   0       NaN  volkswagen   
180      3_reihe  150,000km                  12    benzin       mazda   
226          911    5,000km                  12    benzin     porsche   

    unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
27                NaN  2016-03-27 00:00:00               0        66701   
71                NaN  2016-03-28 00:00:00               0         4552   
80                 ja  2016-03-09 00:00:00               0        66903   
87                NaN  2016-03-29 00:00:00               0        82256   
99               nein  2016-04-05 00:00:00               0        99735   
118                ja  2016-03-12 00:00:00               0        15370   
146               NaN  2016-03-22 00:00:00               0        63069   
167               NaN  2016-04-02 00:00:00               0        64739   
180                ja  2016-03-19 00:00:00               0        30966   
226              nein  2016-03-25 00:00:00               0        44575   

               last_seen  
27   2016-03-27 18:45:01  
71   2016-04-07 01:45:48  
80   2016-03-09 16:43:50  
87   2016-04-06 21:18:15  
99   2016-04-07 12:17:34  
118  2016-03-12 21:44:23  
146  2016-04-01 20:16:38  
167  2016-04-06 19:45:08  
180  2016-03-24 03:17:21  
226  2016-04-05 14:46:39

In [123]:
autos.loc[zero_price_bool].tail(10)

date_crawled                                               name  \
49614  2016-03-09 17:49:39  Verkauf_meine__Audi__A3_.Gegen_tauschen_mit_Di...   
49739  2016-03-17 15:47:34                          Reparaturfaelligen_twingo   
49755  2016-04-02 13:46:23                        Mercedes_Benz_w202_unfaller   
49793  2016-03-05 14:58:49                                  verkaufe_BMW_E_39   
49880  2016-03-30 08:52:57                                E39_528i_an_Bastler   
49884  2016-03-11 13:55:30               Audi_a6_2.5l__Schnaeppchen_nur_heute   
49943  2016-03-16 20:46:08                                         Opel_astra   
49960  2016-03-25 22:51:55            Ford_KA_zu_verschenken_***Reserviert***   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   

      price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
49614    $0     test          NaN               2016    manuell       101   
49739    $0     test          NaN               2016    manuell        58   
49755    $0     test    limousine               1998    manuell       136   
49793    $0  control        kombi               2000  automatik       183   
49880    $0  control          NaN               2017    manuell       193   
49884    $0     test        kombi               1999    manuell       150   
49943    $0  control          NaN               2016    manuell       101   
49960    $0  control   kleinwagen               1999    manuell        60   
49974    $0  control       cabrio               1983    manuell        70   
49984    $0     test          NaN               2000        NaN         0   

          model   odometer  registration_month fuel_type           brand  \
49614        a3  150,000km                   0    benzin            audi   
49739    twingo  150,000km                   0    benzin         renault   
49755  c_klasse  150,000km                   7    benzin   mercedes_benz   
49793       5er  150,000km                   1    benzin             bmw   
49880       5er  150,000km                   4       NaN             bmw   
49884        a6  150,000km                  11    diesel            audi   
49943     astra  150,000km                   8    benzin            opel   
49960        ka  150,000km                   6    benzin            ford   
49974      golf  150,000km                   2    benzin      volkswagen   
49984       NaN  150,000km                   0       NaN  sonstige_autos   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
49614              nein  2016-03-09 00:00:00               0        47051   
49739              nein  2016-03-17 00:00:00               0        27412   
49755               NaN  2016-04-02 00:00:00               0        99441   
49793               NaN  2016-03-05 00:00:00               0        13125   
49880                ja  2016-03-30 00:00:00               0        65468   
49884               NaN  2016-03-11 00:00:00               0        27711   
49943               NaN  2016-03-16 00:00:00               0        89134   
49960               NaN  2016-03-25 00:00:00               0        34355   
49974              nein  2016-03-20 00:00:00               0         8209   
49984               NaN  2016-03-31 00:00:00               0        12103   

                 last_seen  
49614  2016-03-10 14:15:53  
49739  2016-03-17 15:47:34  
49755  2016-04-02 13:46:23  
49793  2016-03-05 18:47:14  
49880  2016-04-07 01:15:27  
49884  2016-03-12 03:17:08  
49943  2016-03-17 19:44:20  
49960  2016-03-25 22:51:55  
49974  2016-03-27 19:48:16  
49984  2016-04-02 19:44:53

Notice the price column.

### Unique value counts for the "number of pictures" column when price is zero

In [124]:
# unique value counts for the number of pictures column when price is $0
autos.loc[zero_price_bool, "nr_of_pictures"].value_counts(dropna=False)

0    1421
Name: nr_of_pictures, dtype: int64

For every single instance in which the car price is zero, the number of pictures taken for that car is also zero. This sticks out as a red flag for obvious reasons (i.e., any reasonable person would not be selling their car for $0 while showing zero pictures of said car). It is possible 
* that these listings are not real (something happened with the web scrape), 
* that these listings were real but taken down (yet the web scrape still picked up on it somehow and this is the result), 
* or that some users have been posting empty listings as a mistake / when trying to make themselves familiar with the website (1421 instances out of 50,000 is about 2.8%).

Nonetheless, we will go ahead and omit this data.

### Drop every row where the price is zero

In [125]:
# drop every row where the price is $0
autos = autos.loc[~zero_price_bool]

autos.describe(include=['O'])

date_crawled         name  price ab_test vehicle_type  gearbox  \
count                 48579        48579  48579   48579        43990    46232   
unique                46895        37481   2356       2            8        2   
top     2016-03-29 23:42:13  Ford_Fiesta   $500    test    limousine  manuell   
freq                      3           76    781   25025        12603    36110   

        model   odometer fuel_type       brand unrepaired_damage  \
count   46116      48579     44544       48579             39472   
unique    245         13         7          40                 2   
top      golf  150,000km    benzin  volkswagen              nein   
freq     3900      31422     29375       10339             34782   

                 ad_created            last_seen  
count                 48579                48579  
unique                   76                38483  
top     2016-04-03 00:00:00  2016-04-07 06:17:27  
freq                   1887                    8

We should also more deeply investigate rows where the number of pictures is zero, just in case it reveals additional anomalies.

## Investigating rows where the number of pictures is zero

In [126]:
zero_pics_bool = autos["nr_of_pictures"] == 0
autos.loc[zero_pics_bool].head(10)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

    price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   
5  $7,900     test          bus               2006  automatik       150   
6    $300     test    limousine               1995    manuell        90   
7  $1,990  control    limousine               1998    manuell        90   
8    $250     test          NaN               2000    manuell         0   
9    $590  control          bus               1997    manuell        90   

     model   odometer  registration_month fuel_type       brand  \
0   andere  150,000km                   3       lpg     peugeot   
1      7er  150,000km                   6    benzin         bmw   
2     golf   70,000km                   7    benzin  volkswagen   
3   fortwo   70,000km                   6    benzin       smart   
4    focus  150,000km                   7    benzin        ford   
5  voyager  150,000km                   4    diesel    chrysler   
6     golf  150,000km                   8    benzin  volkswagen   
7     golf  150,000km                  12    diesel  volkswagen   
8    arosa  150,000km                  10       NaN        seat   
9   megane  150,000km                   7    benzin     renault   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   
3              nein  2016-03-12 00:00:00               0        33729   
4              nein  2016-04-01 00:00:00               0        39218   
5               NaN  2016-03-21 00:00:00               0        22962   
6               NaN  2016-03-20 00:00:00               0        31535   
7              nein  2016-03-16 00:00:00               0        53474   
8              nein  2016-03-22 00:00:00               0         7426   
9              nein  2016-03-16 00:00:00               0        15749   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50  
5  2016-04-06 09:45:21  
6  2016-03-23 02:48:59  
7  2016-04-07 03:17:32  
8  2016-03-26 18:18:10  
9  2016-04-06 10:46:35

In [127]:
autos.loc[zero_pics_bool].tail(10)

date_crawled                                               name  \
49990  2016-03-21 19:54:19         Mercedes_Benz_A_200__BlueEFFICIENCY__Urban   
49991  2016-03-06 15:25:19                                         Kleinwagen   
49992  2016-03-10 19:37:38              Fiat_Grande_Punto_1.4_T_Jet_16V_Sport   
49993  2016-03-15 18:47:35           Audi_A3__1_8l__Silber;_schoenes_Fahrzeug   
49994  2016-03-22 17:36:42  Audi_A6__S6__Avant_4.2_quattro_eventuell_Tausc...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
49990  $17,500     test    limousine               2012    manuell       156   
49991     $500  control          NaN               2016    manuell         0   
49992   $4,800  control   kleinwagen               2009    manuell       120   
49993   $1,650  control   kleinwagen               1997    manuell         0   
49994   $5,000  control        kombi               2001  automatik       299   
49995  $24,900  control    limousine               2011  automatik       239   
49996   $1,980  control       cabrio               1996    manuell        75   
49997  $13,200     test       cabrio               2014  automatik        69   
49998  $22,900  control        kombi               2013    manuell       150   
49999   $1,250  control    limousine               1996    manuell       101   

          model   odometer  registration_month fuel_type          brand  \
49990  a_klasse   30,000km                  12    benzin  mercedes_benz   
49991    twingo  150,000km                   0    benzin        renault   
49992    andere  125,000km                   9       lpg           fiat   
49993       NaN  150,000km                   7    benzin           audi   
49994        a6  150,000km                   1    benzin           audi   
49995        q5  100,000km                   1    diesel           audi   
49996     astra  150,000km                   5    benzin           opel   
49997       500    5,000km                  11    benzin           fiat   
49998        a3   40,000km                  11    diesel           audi   
49999    vectra  150,000km                   1    benzin           opel   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
49990              nein  2016-03-21 00:00:00               0        58239   
49991               NaN  2016-03-06 00:00:00               0        61350   
49992              nein  2016-03-10 00:00:00               0        68642   
49993               NaN  2016-03-15 00:00:00               0        65203   
49994              nein  2016-03-22 00:00:00               0        46537   
49995              nein  2016-03-27 00:00:00               0        82131   
49996              nein  2016-03-28 00:00:00               0        44807   
49997              nein  2016-04-02 00:00:00               0        73430   
49998              nein  2016-03-08 00:00:00               0        35683   
49999              nein  2016-03-13 00:00:00               0        45897   

                 last_seen  
49990  2016-04-06 22:46:57  
49991  2016-03-06 18:24:19  
49992  2016-03-13 01:44:51  
49993  2016-04-06 19:46:53  
49994  2016-04-06 08:16:39  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

In [128]:
autos.loc[zero_pics_bool].describe(include=['O'])

date_crawled         name  price ab_test vehicle_type  gearbox  \
count                 48579        48579  48579   48579        43990    46232   
unique                46895        37481   2356       2            8        2   
top     2016-03-29 23:42:13  Ford_Fiesta   $500    test    limousine  manuell   
freq                      3           76    781   25025        12603    36110   

        model   odometer fuel_type       brand unrepaired_damage  \
count   46116      48579     44544       48579             39472   
unique    245         13         7          40                 2   
top      golf  150,000km    benzin  volkswagen              nein   
freq     3900      31422     29375       10339             34782   

                 ad_created            last_seen  
count                 48579                48579  
unique                   76                38483  
top     2016-04-03 00:00:00  2016-04-07 06:17:27  
freq                   1887                    8

No red flags.

## Converting price to numeric data type

In [129]:
autos["price"] = autos["price"]\
                 .str.replace("$", "")\
                 .str.replace(",", "")\
                 .astype(int)         

# rename column
autos.rename({"price":"price_USD"}, axis=1, inplace=True)

autos.head(2)

date_crawled                                        name  price_USD  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD       5000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik       8500   

   ab_test vehicle_type  registration_year    gearbox  power_PS   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   

    odometer  registration_month fuel_type    brand unrepaired_damage  \
0  150,000km                   3       lpg  peugeot              nein   
1  150,000km                   6    benzin      bmw              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08

## Converting odometer to numeric data type

In [130]:
autos["odometer"] = autos["odometer"]\
                    .str.replace("km", "")\
                    .str.replace(",", "")\
                    .astype(float)
            
# rename column
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

autos.head(2)

date_crawled                                        name  price_USD  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD       5000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik       8500   

   ab_test vehicle_type  registration_year    gearbox  power_PS   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   

   odometer_km  registration_month fuel_type    brand unrepaired_damage  \
0     150000.0                   3       lpg  peugeot              nein   
1     150000.0                   6    benzin      bmw              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08

In [131]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48579 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          48579 non-null object
name                  48579 non-null object
price_USD             48579 non-null int32
ab_test               48579 non-null object
vehicle_type          43990 non-null object
registration_year     48579 non-null int64
gearbox               46232 non-null object
power_PS              48579 non-null int64
model                 46116 non-null object
odometer_km           48579 non-null float64
registration_month    48579 non-null int64
fuel_type             44544 non-null object
brand                 48579 non-null object
unrepaired_damage     39472 non-null object
ad_created            48579 non-null object
nr_of_pictures        48579 non-null int64
postal_code           48579 non-null int64
last_seen             48579 non-null object
dtypes: float64(1), int32(1), int64(5), object(11)
memory usage: 6.9+ MB


# Deeper cleaning of odometer_km

In [132]:
odo = autos["odometer_km"]
odo.unique().shape

(13,)

In [133]:
odo.describe()

count     48579.000000
mean     125766.174685
std       39792.936148
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [134]:
odo.value_counts().sort_index(ascending=True)

5000.0        837
10000.0       253
20000.0       762
30000.0       780
40000.0       816
50000.0      1014
60000.0      1155
70000.0      1217
80000.0      1415
90000.0      1734
100000.0     2116
125000.0     5058
150000.0    31422
Name: odometer_km, dtype: int64

No glaring outliers; everything seems reasonable.

# Deeper cleaning of price

In [135]:
price = autos["price_USD"]
price.unique().shape

(2356,)

In [136]:
price.describe()

count    4.857900e+04
mean     1.012788e+04
std      4.880873e+05
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      1.000000e+08
Name: price_USD, dtype: float64

We have outliers: a car that's selling for \$1, and a car that's selling for $100,000,000.

In [137]:
price_frequencies = price.value_counts().sort_index(ascending=True)
price_frequencies

1           156
2             3
3             1
5             2
8             1
9             1
10            7
11            2
12            3
13            2
14            1
15            2
17            3
18            1
20            4
25            5
29            1
30            7
35            1
40            6
45            4
47            1
49            4
50           49
55            2
59            1
60            9
65            5
66            1
70           10
           ... 
151990        1
155000        1
163500        1
163991        1
169000        1
169999        1
175000        1
180000        1
190000        1
194000        1
197000        1
198000        1
220000        1
250000        1
259000        1
265000        1
295000        1
299000        1
345000        1
350000        1
999990        1
999999        2
1234566       1
1300000       1
3890000       1
10000000      1
11111111      2
12345678      3
27322222      1
99999999      1
Name: price_USD, Length:

Cars being listed as less than \$100 can be explained by the fact that eBay is an auctioning site; therefore, we can expect circumstances where the opening bid is genuinely \$1. When we look at the upper outliers, however, there is a gradual trend up to \$350,000, but after that the numbers start becoming unreasonable (the next value is a 3x jump; we see "1234566" and "12345678"). Therefore we'll cut all values above this upper bound.

In [138]:
autos = autos[autos["price_USD"].between(1, 351000)]
autos["price_USD"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_USD, dtype: float64

Our standard deviation is a lot better now.

# Exploring dates

Five columns in our dataset represet date information:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [139]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

These three columns are currently in string form, and we'll later need to convert them to numerical. For now, we can still take a look at each's date distributions.

In [140]:
(autos['date_crawled']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

**Observations:** The crawler covered roughly an equal amount each day.

In [141]:
(autos['ad_created']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

**Observations:** The oldest ad extends back 10 months. The distribution follows intuition in that the proportion of older ads tends to be less than the proportion of newer ads. (More of the older ads were likely closed/completed before the crawler started).

In [142]:
(autos['last_seen']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

**Observations:** Because the distribution of last seen dates is roughly even, this suggests that cars were sold or ads closed at a roughly even rate. (Once a car is no longer seen, that means it's not being listed and therefore it's been sold or the ad has closed). The spike at the end of the distribution is more likely explained by the fact that the crawler stopped and probably not because there was a genuine spike in sales. 

Now that we've looked at date crawled, ad created, and last seen, let's take a look at the distribution of registration years as well.

In [143]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

**Observations:** Some clear outliers -- 1000, 9999; anything later than 2016 would not make sense. The lower bound cutoff is less clear - it could be anywhere in the 1900s.

## Dealing with incorrect car registration data

In [144]:
# select registration years NOT between 1980 and 2016
(autos.loc[~autos["registration_year"].between(1980, 2016), "registration_year"]
     .value_counts()
     .sort_index())

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
1965      17
1966      22
1967      26
1968      26
1969      19
1970      38
1971      26
1972      33
1973      23
1974      24
1975      18
1976      21
1977      22
1978      44
1979      34
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

If we were generous, we would draw the line at 1910. But it seems more practical for us to draw the line at 1960, the first point where we have double digits. In this sense, we are interpreting the scope of our data to exclude extreme antique cars. 

In [145]:
autos = autos[autos["registration_year"].between(1960, 2016)]
autos["registration_year"].describe()

count    46634.000000
mean      2002.969400
std          6.930433
min       1960.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [146]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1960    0.000493
1961    0.000129
1962    0.000086
1963    0.000172
1964    0.000257
1965    0.000365
1966    0.000472
1967    0.000558
1968    0.000558
1969    0.000407
1970    0.000815
1971    0.000558
1972    0.000708
1973    0.000493
1974    0.000515
1975    0.000386
1976    0.000450
1977    0.000472
1978    0.000944
1979    0.000729
1980    0.001823
1981    0.000600
1982    0.000879
1983    0.001094
1984    0.001094
1985    0.002037
1986    0.001544
1987    0.001544
1988    0.002895
1989    0.003731
1990    0.007441
1991    0.007269
1992    0.007934
1993    0.009114
1994    0.013488
1995    0.026311
1996    0.029442
1997    0.041836
1998    0.050671
1999    0.062122
2000    0.067676
2001    0.056525
2002    0.053309
2003    0.057876
2004    0.057962
2005    0.062958
2006    0.057254
2007    0.048827
2008    0.047498
2009    0.044710
2010    0.034074
2011    0.034803
2012    0.028091
2013    0.017219
2014    0.014217
2015    0.008406
2016    0.026161
Name: registration_year, dtype:

**Observations:** Most of the cars were registered between 1995 and 2012, with the greatest proportion coming from 2000. 

There's an interesting spike in 2016 that defies the steady decline in proportion from 2005 to 2015 (the steady decline is intuitive: cars registered recently have been owned by their current owner for less time, so those owners are not ready to put their cars up for sale). What explains the 2016 spike? One possibility:

- Cars depreciate most quickly in their [first year](https://www.carfax.com/blog/car-depreciation), so if someone doesn't like the car they just bought, they're incentivized to get rid of it ASAP. This may not a comprehensive explanation because it is certainly not the case that every car registered in 2016 has only had one previous owner.

## Conversion of dates to numeric

Recall that we had:

In [154]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We'll convert these so that a value such as `2016-03-21` becomes the integer `20160321`.

In [155]:
for column in ['date_crawled','ad_created','last_seen']:
    autos[column] = autos[column].str[:10].str.replace("-", "").astype(int)
    
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled  ad_created  last_seen
0      20160326    20160326   20160406
1      20160404    20160404   20160406
2      20160326    20160326   20160406
3      20160312    20160312   20160315
4      20160401    20160401   20160401

# Analysis of prices by top 5 brands

In [147]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

These are all the different brands in the dataset.

In [148]:
top_5 = autos["brand"].value_counts().sort_values(ascending=False).head(5)
top_5

volkswagen       9860
bmw              5136
opel             5018
mercedes_benz    4495
audi             4041
Name: brand, dtype: int64

These are the top 5 brands.

Admittedly, we are not yet controlling for mileage, but we can still find generic trends in average price by brand.

In [149]:
avg_price_by_brand = {}
for top_brand in top_5.index:
    avg = autos.loc[autos["brand"] == top_brand, "price_USD"].mean()
    avg_price_by_brand[top_brand] = avg
    
avg_price_by_brand

{'audi': 9336.687453600594,
 'bmw': 8332.203855140187,
 'mercedes_benz': 8565.086095661847,
 'opel': 2974.688122758071,
 'volkswagen': 5398.871196754564}

**Observation:** We can see that opel is on the lower end (least expensive), volkswagen is on the medium-level, and bmw, mercedez benz, and audi are on the higher end (most expensive).

## Considering mileage (odometer) in our analysis

We will put average mileage for each of the top 5 brands side-by-side with the average pricing to see to what extent it interferes with such brand pricing.

In [150]:
avg_mileage_by_brand = {}
for top_brand in top_5.index:
    avg = autos.loc[autos["brand"] == top_brand, "odometer_km"].mean()
    avg_mileage_by_brand[top_brand] = avg
    
avg_mileage_by_brand

{'audi': 129157.38678544914,
 'bmw': 132597.35202492212,
 'mercedes_benz': 130919.91101223581,
 'opel': 129342.3674770825,
 'volkswagen': 128707.9107505071}

In [153]:
bmp_series = pd.Series(avg_price_by_brand) # brand mean price
df = pd.DataFrame(bmp_series, columns=['mean_price_USD'])

bmm_series = pd.Series(avg_mileage_by_brand) # brand mean mileage
df['mean_mileage_km'] = bmm_series

df

mean_price_USD  mean_mileage_km
audi              9336.687454    129157.386785
bmw               8332.203855    132597.352025
mercedes_benz     8565.086096    130919.911012
opel              2974.688123    129342.367477
volkswagen        5398.871197    128707.910751

**Observations:** We can conclude that mileage is not interfering dramatically with our brand pricing. For instance, Opel and Volkswagen remain cheaper even though their mean mileages are relatively lower than the more expensive cars  BMW & Mercedez Benz (though all the mileages are more or less even). It's highly unlikely that the extra ~1500 km on the Mercedez is the main cause for making it \$800 cheaper than the Audi when the relative difference is 1% (1500 / 129,000).